### PR TITLE
Ffwd Output: Extend filtering functionality

### PR DESCRIPTION
**Pull request by nezdolik**
Proposed change makes possible to specify per-plugin filter for ffwd output. Following output plugins were patched and smoke tested: signalfx, http, debug. 

In future we should patch all remaining output plugins and all input plugins.

### CODE REVIEW

File: api/src/main/java/com/spotify/ffwd/output/OutputPlugin.java [hunk 1]
```diff
@@ -18,12 +18,61 @@
...
+        final Key<? extends PluginSink> input, final Key<PluginSink> output
+    ) {
+        return new PrivateModule() {
+            @Override
+            protected void configure() {
+                Key<PluginSink> sinkKey = (Key<PluginSink>) input;
+
+                if (flushInterval.isPresent() || BatchedPluginSink.class.isAssignableFrom(
+                    sinkKey.getTypeLiteral().getRawType())) {
+                    final Key<PluginSink> nextKey = Key.get(PluginSink.class, Names.named("flushing"));
```
Comment (telecatser) [inline api/src/main/java/com/spotify/ffwd/output/OutputPlugin.java]: Maybe rename this variable `nextKey`? Like `flushingPluginKey`?

File: api/src/main/java/com/spotify/ffwd/output/OutputPlugin.java [hunk 2]
```diff
@@ -18,12 +18,61 @@
...
+                    install(new OutputDelegatingModule<>(sinkKey, nextKey,
+                        new FlushingPluginSink(flushInterval.get())));
+                    sinkKey = nextKey;
+                }
+
+                if (filter.isPresent()) {
+                    final Key<PluginSink> next = Key.get(PluginSink.class, Names.named("filtered"));
+                    install(new OutputDelegatingModule<>(sinkKey, next,
+                        new FilteringPluginSink(filter.get())));
+                    sinkKey = next;
```
Comment (telecatser) [inline api/src/main/java/com/spotify/ffwd/output/OutputPlugin.java]: And here, replace `next` with something?

File: api/src/main/java/com/spotify/ffwd/Plugin.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.spotify.ffwd.filter.Filter;
+import com.spotify.ffwd.filter.TrueFilter;
+import java.util.Optional;
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+@Data
+public abstract class Plugin {
```
Comment (udoprog) [inline api/src/main/java/com/spotify/ffwd/Plugin.java]: This doesn't actually seem like it needs to be an abstract class, or even exist at all?

It has no state which couldn't be stored in OutputPlugin.
If we do end up duplicating some state, that's not the end of the world either.

File: api/src/main/java/com/spotify/ffwd/Plugin.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    protected final String id;
+    protected final Optional<Filter> filter;
+
+    protected Plugin(final String id, final Optional<Filter> filter) {
+        this.id = id;
+        this.filter = filter.isPresent() ? filter : Optional.of(new TrueFilter());
```
Comment (udoprog) [inline api/src/main/java/com/spotify/ffwd/Plugin.java]: `filter` field should be empty if not specified to prevent the delegator being set up in `OutputPlugin`.

File: api/src/main/java/com/spotify/ffwd/Plugin.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import com.spotify.ffwd.filter.Filter;
+import com.spotify.ffwd.filter.TrueFilter;
+import java.util.Optional;
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+@Data
+public abstract class Plugin {
+    protected final String id;
```
Comment (udoprog) [inline api/src/main/java/com/spotify/ffwd/Plugin.java]: This id doesn't appear to be used.

File: api/src/main/java/com/spotify/ffwd/output/FilteringPluginSink.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    protected PluginSink sink;
+
+    protected final Filter filter;
+
+    public FilteringPluginSink(final Filter filter) {
+        this.filter = filter;
+    }
+
+    @Override
+    public void init() {
```
Comment (udoprog) [inline api/src/main/java/com/spotify/ffwd/output/FilteringPluginSink.java:38]: This needs to call the delegate `sink`.

File: core/src/main/java/com/spotify/ffwd/AgentCore.java [hunk 7]
```diff
@@ -213,7 +213,7 @@ private Injector setupEarlyInjector() throws Exception {
                 filters.put("key", new MatchKey.Deserializer());
                 filters.put("=", new MatchTag.Deserializer());
                 filters.put("true", new TrueFilter.Deserializer());
-                filters.put("false", new TrueFilter.Deserializer());
+                filters.put("false", new NotFilter.Deserializer());
```
Comment (udoprog) [inline core/src/main/java/com/spotify/ffwd/AgentCore.java]: Wrong filter, should be `FalseFilter`.

File: core/src/main/java/com/spotify/ffwd/AgentCore.java [hunk 8]
```diff
@@ -223,7 +223,7 @@ private Injector setupEarlyInjector() throws Exception {
 
             @Singleton
             @Provides
-            @Named("application/yaml+config")
+            @Named("config")
```
Comment (udoprog) [inline core/src/main/java/com/spotify/ffwd/AgentCore.java:227]: unrelated change?

File: core/src/main/java/com/spotify/ffwd/debug/DebugOutputPlugin.java [hunk 9]
```diff
@@ -19,33 +19,38 @@
...
             protected void configure() {
-                if (flushInterval != null) {
-                    bind(BatchedPluginSink.class).to(DebugPluginSink.class);
-                    bind(key).toInstance(new FlushingPluginSink(flushInterval));
-                } else {
-                    bind(key).to(DebugPluginSink.class).in(Scopes.SINGLETON);
-                }
-
+                bind(BatchedPluginSink.class).to(DebugPluginSink.class);
+                Key<PluginSink> sinkKey = Key.get(PluginSink.class, Names.named("debugSinkSink"));
```
Comment (udoprog) [inline core/src/main/java/com/spotify/ffwd/debug/DebugOutputPlugin.java:51]: This could probably be automated in `wrapPluginSink`?

File: modules/riemann/src/main/java/com/spotify/ffwd/riemann/RiemannOutputPlugin.java [hunk 10]
```diff
@@ -36,25 +35,24 @@
...
-        @JsonProperty("protocol") ProtocolFactory protocol, @JsonProperty("retry") RetryPolicy retry
+        @JsonProperty("flushInterval") Optional<Long> flushInterval,
+        @JsonProperty("protocol") ProtocolFactory protocol,
+        @JsonProperty("retry") RetryPolicy retry, @JsonProperty("filter") Optional<Filter> filter
+
     ) {
-        this.filter = Optional.ofNullable(filter).orElseGet(TrueFilter::new);
-        this.flushInterval = Optional.ofNullable(flushInterval).orElse(DEFAULT_FLUSH_INTERVAL);
+        super("", filter,
+            flushInterval.isPresent() ? flushInterval : Optional.of(DEFAULT_FLUSH_INTERVAL));
```
Comment (udoprog) [inline modules/riemann/src/main/java/com/spotify/ffwd/riemann/RiemannOutputPlugin.java]: This seems wrong. `DEFAULT_FLUSH_INTERVAL == 0` has the same effect as being absent.

Somewhat unrelated to what you did, but since you are mucking around, you can fix it.

**Review by udoprog — CHANGES_REQUESTED** [review]
In general, it doesn't seem like `Plugin` needs to have an ID at all, or at least I'm not seeing it.

Don't forget to add license headers to all new files.

**Comment by nezdolik** [comment]
Applied review comments (except '_This could probably be automated in wrapPluginSink?_')

File: api/src/main/java/com/spotify/ffwd/output/OutputPlugin.java [hunk 11]
```diff
@@ -20,10 +20,60 @@
...
+    }
+
+    public OutputPlugin(
+        final Optional<Filter> filter, final Optional<Long> flushInterval
+    ) {
+        this.filter = filter;
+        this.flushInterval = flushInterval;
+    }
+
+    protected Module wrapPluginSink(
```
Comment (udoprog) [inline api/src/main/java/com/spotify/ffwd/output/OutputPlugin.java]: This method really badly needs documentation.

**Review by udoprog — CHANGES_REQUESTED** [review]

**Comment by nezdolik** [comment]
added jdoc for com.spotify.ffwd.output.OutputPlugin#wrapPluginSink

**Comment by nezdolik** [comment]
added headers

**Comment by nezdolik** [comment]
added [FilteringPluginSinkTest](https://github.com/spotify/ffwd/pull/45/files#diff-a945793520c571e72f94f889e8411af6)

**Review by udoprog — APPROVED** [review]
